// ngx-highlight-js/bundles/ngx-highlight-js.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(require('@angular/core'),require('@angular/forms'),exports, require('@angular/core'), require('@angular/common'), require('@angular/forms')) :
    typeof define === 'function' && define.amd ? define('ngx-highlight-js', ['@angular/core','@angular/forms','exports', '@angular/core', '@angular/common', '@angular/forms'], factory) :
    (global = global || self, factory(global.ng.core,global.ng.forms,global['ngx-highlight-js'] = {}, global.ng.core, global.ng.common, global.ng.forms));
}(this, (function (ɵngcc0,ɵngcc1,exports, core, common, forms) { 'use strict';

    var HighlightJsDirective = /** @class */ (function () {
        function HighlightJsDirective(el, ngModel, doc) {
            this.el = el;
            this.ngModel = ngModel;
            this.doc = doc;
            this.lang = 'html';
        }
        HighlightJsDirective.prototype.escapeHTML = function (str) {
            return (str || '')
                .replace(/&/g, '&amp;')
                .replace(/</g, '&lt;')
                .replace(/>/g, '&gt;')
                .replace(/"/g, '&quot;')
                .replace(/'/g, '&apos;');
        };
        HighlightJsDirective.prototype.init = function () {
            this.destroy();
            this.codeEl = this.doc.createElement('pre');
            if (this.lang) {
                this.codeEl.className = this.lang;
            }
            this.codeEl.innerHTML =
                this.code || '' + this.el.nativeElement.innerHTML.trim();
            this.parentEl = this.el.nativeElement.parentNode;
            this.parentEl.insertBefore(this.codeEl, this.el.nativeElement.nextSibling);
            hljs.configure(Object.assign({}, this.options));
            hljs.highlightBlock(this.codeEl);
        };
        HighlightJsDirective.prototype.destroy = function () {
            if (this.codeEl) {
                this.parentEl.removeChild(this.codeEl);
                this.codeEl = null;
            }
        };
        HighlightJsDirective.prototype.ngOnInit = function () {
            this.init();
        };
        HighlightJsDirective.prototype.ngAfterViewInit = function () {
            var _this = this;
            if (this.ngModel) {
                this.modelValue$ = this.ngModel.valueChanges.subscribe(function (res) {
                    _this.code = _this.escapeHTML(res);
                    _this.init();
                });
            }
            else {
                this.initMutation();
            }
        };
        HighlightJsDirective.prototype.ngOnDestroy = function () {
            this.destroy();
            this.destroyMutation();
            if (this.modelValue$) {
                this.modelValue$.unsubscribe();
            }
        };
        HighlightJsDirective.prototype.initMutation = function () {
            if (typeof MutationObserver === 'undefined') {
                return;
            }
            this.observer = new MutationObserver(this.init.bind(this));
            this.observer.observe(this.el.nativeElement, {
                characterData: true,
                childList: true,
                subtree: true,
            });
        };
        HighlightJsDirective.prototype.destroyMutation = function () {
            if (!this.observer) {
                return;
            }
            this.observer.disconnect();
        };
HighlightJsDirective.ɵfac = function HighlightJsDirective_Factory(t) { return new (t || HighlightJsDirective)(ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ElementRef), ɵngcc0.ɵɵdirectiveInject(ɵngcc1.NgModel, 8), ɵngcc0.ɵɵdirectiveInject(common.DOCUMENT)); };
HighlightJsDirective.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: HighlightJsDirective, selectors: [["", "highlight-js", ""]], hostAttrs: [2, "display", "none"], inputs: { lang: "lang", options: "options", code: "code" }, exportAs: ["highlightJs"] });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(HighlightJsDirective, [{
        type: core.Directive,
        args: [{
                selector: '[highlight-js]',
                host: {
                    style: 'display:none;'
                },
                exportAs: 'highlightJs'
            }]
    }], function () { return [{ type: ɵngcc0.ElementRef }, { type: ɵngcc1.NgModel, decorators: [{
                type: core.Optional
            }] }, { type: undefined, decorators: [{
                type: core.Inject,
                args: [common.DOCUMENT]
            }] }]; }, { lang: [{
            type: core.Input
        }], options: [{
            type: core.Input
        }], code: [{
            type: core.Input
        }] }); })();
        return HighlightJsDirective;
    }());
    HighlightJsDirective.ctorParameters = function () { return [
        { type: core.ElementRef },
        { type: forms.NgModel, decorators: [{ type: core.Optional }] },
        { type: undefined, decorators: [{ type: core.Inject, args: [common.DOCUMENT,] }] }
    ]; };
    HighlightJsDirective.propDecorators = {
        options: [{ type: core.Input }],
        lang: [{ type: core.Input }],
        code: [{ type: core.Input }]
    };

    var HighlightJsModule = /** @class */ (function () {
        function HighlightJsModule() {
        }
HighlightJsModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: HighlightJsModule });
HighlightJsModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function HighlightJsModule_Factory(t) { return new (t || HighlightJsModule)(); }, imports: [[forms.FormsModule]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(HighlightJsModule, { declarations: [HighlightJsDirective], imports: [ɵngcc1.FormsModule], exports: [HighlightJsDirective] }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(HighlightJsModule, [{
        type: core.NgModule,
        args: [{
                imports: [forms.FormsModule],
                declarations: [HighlightJsDirective],
                exports: [HighlightJsDirective]
            }]
    }], function () { return []; }, null); })();
        return HighlightJsModule;
    }());

    /**
     * Generated bundle index. Do not edit.
     */

    exports.HighlightJsDirective = HighlightJsDirective;
    exports.HighlightJsModule = HighlightJsModule;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
