// @sweetalert2/ngx-sweetalert2/esm2015/lib/sweetalert2-loader.service.js
import { __decorate, __param } from "tslib";
import { Inject, Injectable } from '@angular/core';
import { swalProviderToken } from './di';
import * as ɵngcc0 from '@angular/core';
let SweetAlert2LoaderService = class SweetAlert2LoaderService {
    // Using any because Angular metadata generator does not understand a pure TS type here
    constructor(swalProvider) {
        this.swalProvider = swalProvider;
    }
    get swal() {
        if (!this.swalPromiseCache) {
            this.preloadSweetAlertLibrary();
        }
        return this.swalPromiseCache;
    }
    preloadSweetAlertLibrary() {
        if (this.swalPromiseCache)
            return;
        const libPromise = isLoader(this.swalProvider)
            ? this.swalProvider()
            : Promise.resolve(this.swalProvider);
        this.swalPromiseCache = libPromise.then(value => isDefaultExport(value) ? value : value.default);
        function isLoader(value) {
            return typeof value === 'function' && value.version === undefined;
        }
        function isDefaultExport(value) {
            return typeof value === 'function';
        }
    }
};
SweetAlert2LoaderService.ɵfac = function SweetAlert2LoaderService_Factory(t) { return new (t || SweetAlert2LoaderService)(ɵngcc0.ɵɵinject(swalProviderToken)); };
SweetAlert2LoaderService.ɵprov = ɵngcc0.ɵɵdefineInjectable({ token: SweetAlert2LoaderService, factory: function (t) { return SweetAlert2LoaderService.ɵfac(t); } });
SweetAlert2LoaderService.ctorParameters = () => [
    { type: undefined, decorators: [{ type: Inject, args: [swalProviderToken,] }] }
];
SweetAlert2LoaderService = __decorate([ __param(0, Inject(swalProviderToken))
], SweetAlert2LoaderService);
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(SweetAlert2LoaderService, [{
        type: Injectable
    }], function () { return [{ type: undefined, decorators: [{
                type: Inject,
                args: [swalProviderToken]
            }] }]; }, null); })();
export { SweetAlert2LoaderService };
